app/models/assignment.rb

```diff
@@ -600,6 +600,20 @@ def summary_json(user)
       numMarked: self.get_num_marked(user.admin? ? nil : user.id) }
   end
 
+  # Generate a JSON summary of the test results associated with an assignment
+  def summary_test_result_json(user)
+    return {} unless user.admin? || user.ta?
+    {
+      submission_test_results: self.submissions.map do |submission|
+        {
+          group_name: submission.grouping.group.group_name,
+          submission_date: submission.created_at,
+          test_results: [*submission.test_group_results.order('created_at').last]
```

Comment:
This is a good start, but needs some refinement.

1. I want to see the `TestResult` objects here, in addition to the fields from `TestGroupResult`.
2. An assignment can have more than one `TestGroup`, so what you really want is the last `TestGroupResult` *for each `TestGroup`*. That's a bit trickier to do. One possibility is using fancier SQL to get all of the "most recent test group results" for an assignment (across all submissions), and then divide those up by submission id. We do something similar here: https://github.com/MarkUsProject/Markus/blob/438b7c36d69c9cdd293aaea77c048fe601e3962b/app/models/assignment.rb#L793-L802.
3. Right now this code has an [N+1 query problem](https://guides.rubyonrails.org/active_record_querying.html#eager-loading-associations), since the test group results are queried once per submission. If you do what I suggested in point 2 above, that'll fix the issue for `TestGroupResult`, but not `TestResult`. For the latter, one option is to load all of the relevant `TestResult` objects and then group them by `test_group_result_id`.
4. We don't need every field in these objects. You can use the `pluck_to_hash` query method to grab only certain fields from the database. 

    - For `TestGroupResult`, we need the `name` of the associated `TestGroup`, `marks_earned`, `marks_total`, `extra_info`, and `error_type`.
    - For `TestResult`, we need `name`, `status`, `marks_earned`, `marks_total`, and `output`.